### PR TITLE
matrix exp(A) returns real-valued adjoint for real-valued A

### PR DESCRIPTION
**Pull request by sdewaele**
The current adjoint for the matrix `exp(A)` where `A` is real-valued can return a complex-valued matrix. The complex components are only small numerical errors, close to machine precision. However, the fact that it is complex-valued leads to issues in the adjoint computation, see below for an example. This PR converts the return value to a real-valued array for real-valued arguments.

BTW - This issue has some similarities to #402, in that
1) Having indexing in the test code revealed the issue. For that reason, it may be an idea to add a `gradcheck_getindex`, to complement the current `gradcheck`? Or a test of compatibility of the return value of the adjoint with the input type?

2) An alternative fix would be to allow the adjoint of the indexed array to be complex-valued. This is similar to allowing unconstrainted matrices as adjoints in #402. This would also require changes to the `getindex` adjoint.

```julia
# Note: A randomly generated 8 × 8 also fails (almost?) always. Using this confirmed example to be sure.
A = [ 0.0    1.0    0.0
      0.0    0.0    1.0
     -4.34 -18.31  -0.43]
x = [1.0]

f(A,x) = exp(A*x[1])
Y,Bf = Zygote.pullback(f,A,x)
Ȳ = ones(3,3)
Ā,x̄ = Bf(Ȳ) # fails
```
Error:
```julia
ERROR: InexactError: Float64(14.999150171071268 + 9.678155693021492e-16im)
Stacktrace:
 [1] Type at .\complex.jl:37 [inlined]
 [2] convert at .\number.jl:7 [inlined]
 [3] setindex!(::Array{Float64,1}, ::Complex{Float64}, ::Int64) at .\array.jl:767
 [4] #980 at C:\Users\user\.julia\packages\Zygote\ycnjm\src\lib\array.jl:32 [inlined]
 [5] #2619#back at C:\Users\user\.julia\packages\ZygoteRules\6nssF\src\adjoint.jl:49 [inlined]
 [6] literal_getindex at C:\Users\user\.julia\packages\Zygote\ycnjm\src\lib\lib.jl:77 [inlined]
 [7] (::typeof(∂(Zygote.literal_getindex)))(::Complex{Float64}) at C:\Users\user\.julia\packages\Zygote\ycnjm\src\compiler\interface2.jl:0
 [8] f at .\REPL[7]:1 [inlined]
 [9] (::typeof(∂(f)))(::Array{Float64,2}) at C:\Users\user\.julia\packages\Zygote\ycnjm\src\compiler\interface2.jl:0
 [10] (::getfield(Zygote, Symbol("##28#29")){typeof(∂(f))})(::Array{Float64,2}) at C:\Users\user\.julia\packages\Zygote\ycnjm\src\compiler\interface.jl:38
 [11] top-level scope at none:0
```

### CODE REVIEW

**Comment by mcabbott** [comment]
One issue here is whether real parameters should by default be allowed to return complex gradients, as is the current behaviour, see #342 for a discussion. 

The other is that, given this behaviour, the gradient of `getindex` should promote as necessary. This is precisely #376, which is fixed by #256.

File: src/lib/array.jl [hunk 1]
```diff
@@ -485,7 +485,8 @@ end
   X = _pairdiffquotmat(exp, n, w, ew, ew, ew)
   V = E.vectors
   VF = factorize(V)
-  Ā = (V * ((VF \ F̄' * V) .* X) / VF)'
+  Āc = (V * ((VF \ F̄' * V) .* X) / VF)'
+  Ā = eltype(A) <: Real ? real(Āc) : Āc
```
Comment (mcabbott) [inline src/lib/array.jl]:
```suggestion
  Ā = eltype(A) <: Real && eltype(F̄) <: Real ? real(Āc) : Āc
```

**Comment by sdewaele** [comment]
Thanks for your comments! I will update the PR accordingly, and fix the test issue.

File: src/lib/array.jl [hunk 2]
```diff
@@ -485,7 +485,8 @@ end
   X = _pairdiffquotmat(exp, n, w, ew, ew, ew)
   V = E.vectors
   VF = factorize(V)
-  Ā = (V * ((VF \ F̄' * V) .* X) / VF)'
+  Āc = (V * ((VF \ F̄' * V) .* X) / VF)'
+  Ā = eltype(A) <: Real && eltype(F̄) <: Real ? real(Āc) : Āc
```
Comment (MikeInnes) [inline src/lib/array.jl]: Would `isreal(A) && isreal(F)` work?

Comment (sdewaele) [reply src/lib/array.jl]: Yes. I added it. Thanks!

**Comment by sdewaele** [comment]
The check [failed](https://travis-ci.org/FluxML/Zygote.jl/jobs/625158623#L254) for Julia 1.3 with error 
```
No output has been received in the last 10m0s,
this potentially indicates a stalled build or something wrong with the build itself.
```
I am not sure why this is happening; when I test on my machine with Julia 1.3 running the tests takes 5 minutes. Versioninfo:
```
julia>  versioninfo()
Julia Version 1.3.0
Commit 46ce4d7933 (2019-11-26 06:09 UTC)
Platform Info:
  OS: Windows (x86_64-w64-mingw32)
  CPU: Intel(R) Core(TM) i7-7500U CPU @ 2.70GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-6.0.1 (ORCJIT, skylake)
```

**Comment by MikeInnes** [comment]
Travis probably just has a slow CPU. Bors should be fine:

bors r+

**Pull request merged**